rosserial_client/test/subscriber_test.cpp

```diff
@@ -0,0 +1,55 @@
+#include "ros_lib/ros/subscriber.h"
+#include <gtest/gtest.h>
+
+
+bool callbackCalled;
+unsigned char buffer[1];
+
+class DummyMsg
+{
+public:
+  int serialize(unsigned char *outbuffer) const { return 0; }
+  int deserialize(unsigned char *inbuffer) { return 0; }
+  const char * getType() { return ""; }
+  const char * getMD5() { return ""; }
+};
+
+class DummyClass
+{
+public:
+  static void staticCallback(const DummyMsg& msg)
+  {
+    callbackCalled = true;
+  }
+  void memberCallback(const DummyMsg& msg)
+  {
+    callbackCalled = true;
+  }
+};
+
+
+TEST(TestSubscriber, testStaticCallback)
+{
+  ros::Subscriber<DummyMsg> sub("topic_name", &DummyClass::staticCallback);
+
+  callbackCalled = false;
+  sub.callback(buffer);
+  ASSERT_TRUE(callbackCalled);
+}
+
+TEST(TestSubscriber, testMemberCallback)
+{
+  DummyClass cl;
+  ros::Subscriber<DummyMsg, DummyClass> sub("topic_name", &DummyClass::memberCallback, &cl);
```

Comment:
Types can't be inferred for constructors, only regular functions, but to do this as a regular function, the Subscriber instance would have to be heap-allocated.